sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -121,37 +113,37 @@ internal class LazyLoadingReadOnlyStream<TRequestConditions, TProperties> : Stre
         /// </summary>
         private readonly DownloadInternalAsync _downloadInternalFunc;
 
-        /// <summary>
-        /// Function to create RequestConditions.
-        /// </summary>
-        private readonly CreateRequestConditions _createRequestConditionsFunc;
-
         /// <summary>
         /// Function to get properties.
         /// </summary>
         private readonly GetPropertiesAsync _getPropertiesInternalFunc;
 
         public LazyLoadingReadOnlyStream(
             DownloadInternalAsync downloadInternalFunc,
-            CreateRequestConditions createRequestConditionsFunc,
             GetPropertiesAsync getPropertiesFunc,
+            bool allowModifications,
+            ETag? expectedETag,
             long initialLenght,
             long position = 0,
             int? bufferSize = default,
             TRequestConditions requestConditions = default)
         {
             _downloadInternalFunc = downloadInternalFunc;
-            _createRequestConditionsFunc = createRequestConditionsFunc;
             _getPropertiesInternalFunc = getPropertiesFunc;
             _position = position;
             _bufferSize = bufferSize ?? Constants.DefaultStreamingDownloadSize;
             _buffer = ArrayPool<byte>.Shared.Rent(_bufferSize);
+            _allowBlobModifications = allowModifications;
+            _expectedETag = expectedETag;
             _bufferPosition = 0;
             _bufferLength = 0;
             _requestConditions = requestConditions;
             _length = initialLenght;
-            _allowBlobModifications = !(_requestConditions == null && _createRequestConditionsFunc != null);
             _bufferInvalidated = false;
+            if (!_allowBlobModifications && !expectedETag.HasValue)
```

Comment:
What does it mean to pass `allowBlobModifications` and a `requestConditions` with an ETag at the same time?  It doesn't look like that's disallowed here or below?

Reply:
This is a bit bodged now. I added this piece when I realized what Shares logic needs to actually work as expected (and that they won't let me pass etag through their ShareRequestConditions...).

I'll be looking for a better way of doing this. But what we want is:
- use etag in blob conditions for blobs since they support that.
- validate etag after range read call in shares (won't hurt to do this for blobs).

The fact that ShareRequestConditions don't carry IfMatch is really a cause of this ugliness. Instead of doing this we could bring back a delegate that creates requestconditions from etag and leave it "empty" for shares and remove the "requestConditions" from this consturctor.

Reply:
Could we either:
- Take a separate parameter to indicate you're in Shares mode?
- Move all the conditional header logic to virtuals and then derive a SharesLazyLoadingReadOnlyStream?